### PR TITLE
compaction_manager: compaction_disabled: return true if not in compaction_state

### PR DESCRIPTION
**Pull request by bhalevy**
When a compaction_group is removed via `compaction_manager::remove`, it is erase from `_compaction_state`, and therefore compaction is definitely not enabled on it.

This triggers an internal error if tablets are cleaned up during drop/truncate, which checks that compaction is disabled in all compaction groups.

Note that the callers of `compaction_disabled` aren't really interested in compaction being actively disabled on the compaction_group, but rather if it's enabled or not. A follow-up patch can be consider to reverse the logic and expose `compaction_enabled` rather than `compaction_disabled`.

Fixes scylladb/scylladb#20060

* Needs backport to all branches that support tablets

### CODE REVIEW

File: service/topology_coordinator.cc [hunk 1]
```diff
@@ -1173,7 +1173,10 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                 case locator::tablet_transition_stage::streaming:
                     if (drain) {
                         utils::get_local_injector().inject("stream_tablet_fail_on_drain",
-                                        [] { throw std::runtime_error("stream_tablet failed due to error injection"); });
+                                        [] {
+                                            rtlogger.info("Injecting stream_tablet_fail_on_drain");
+                                            throw std::runtime_error("stream_tablet failed due to error injection");
```
Comment (raphaelsc) [inline service/topology_coordinator.cc:1178]: Is this some left over from a test you wrote but didn't push?

File: service/storage_service.cc [hunk 2]
```diff
@@ -5930,10 +5930,11 @@ future<> storage_service::do_tablet_operation(locator::global_tablet_id tablet,
         _tablet_ops.erase(tablet);
     });
 
+    rtlogger.info("{} for tablet migration of {} starting", op_name, tablet);
     try {
         co_await op(guard);
         p.set_value();
-        rtlogger.debug("{} for tablet migration of {} successful", op_name, tablet);
+        rtlogger.info("{} for tablet migration of {} successful", op_name, tablet);
     } catch (...) {
```
Comment (raphaelsc) [inline service/storage_service.cc:5938]: I think you possibly forgot to undo these changes before pushing.

File: compaction/compaction_manager.cc [hunk 3]
```diff
@@ -2243,7 +2245,15 @@ bool compaction_manager::has_table_ongoing_compaction(const table_state& t) cons
...
-    return _compaction_state.contains(&t) && _compaction_state.at(&t).compaction_disabled();
+    if (auto it = _compaction_state.find(&t); it != _compaction_state.end()) {
+        return it->second.compaction_disabled();
+    } else {
+        cmlog.debug("compaction_disabled: {}:{} not in compaction_state", t.schema()->id(), t.get_group_id());
+        // Compaction is not strictly disabled, but it is not enabled either.
+        // The callers actually care about if it's enabled or not, not about the actual state of
+        // compaction_state::compaction_disabled()
+        return true;
+    }
```
Comment (raphaelsc) [inline compaction/compaction_manager.cc:2256]: it's essentially disabled indeed. change looks good to me.

File: service/storage_service.cc (hunk 2 above)
Comment (bhalevy) [reply service/storage_service.cc:5938]: not really.
Ideally, I should have them in a separate patch and explain why.
This information is rare enough not to spam the log, and important enough to be logged at info rather than debug level, to provide the visibility to figure out what's going on in the system around tablet migrations.

File: service/topology_coordinator.cc (hunk 1 above)
Comment (bhalevy) [reply service/topology_coordinator.cc:1178]: The test is preexisting.
It's important to log that an error is being injected to be able to match the node log with the test log.

File: service/storage_service.cc (hunk 2 above)
Comment (raphaelsc) [reply service/storage_service.cc:5938]: A possible spam scenario is high smp count and decommissions. storage_service::stream_tablet() which invokes do_tablet_op() happens to run on destination.

**Comment by bhalevy** [comment]
In v2 ([1c55747](https://github.com/scylladb/scylladb/commit/1c5574763739aad06caece249a76560eb7b891d5)):
- left only the gist of the fix.
- will consider additional logging as a follow up

**Review by raphaelsc — APPROVED** [review]
